keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+
+### Goals
+
+* Provide structured logging interface in K8s codebase
+* Standardize commonly used logging keys (e.g. referencing k8s object, http request fields)
+* Propose sound strategy for migration ensuring its completeness
+* Make implementation of context based solutions like logs correlation or tracing within Kubernetes easier.
+
+### Non-Goals
+
+* We are **not** changing logging pipeline in K8s
+* We are **not** standardizing some specific log messages
+* We are **not** tackling logging keys lifecycle (should be addressed by separate KEP when new API is close to GA)
+
+## Proposal
+
+This KEP proposes introducing a structured logging into Kubernetes control-plane components. As part of this proposal we will introduce a new logging API that will replace klog. We will introduce a Kubernetes logging schema which will provide a way to define a common set of logging keys and ensure their consistent usage. To provide immediate benefit for adapting a new API we will propose a new logging format that will simplify ingestion of structured data into third party logging solutions. As part of changes in kubernetes logging we propose to tightly couple usage of logging and golang context.
+
+### Logging Interface
+ 
+When choosing new structured logging API following criteria has been used:
+* Selected API should provide abstract structured logging interface so that it would be easy to inject different implementations for the same API.
+* Selected API should be minimalistic so that it should be easy to reimplement, wrap or to mock it.
+* Applicability of the selected API should not be limited to the Kubernetes control plane but it should be possible to use the same API in other components running on Kubernetes clusters which are also implemented in go.
+* Selected API should support structured logging by providing a way to define a set of key-value pairs for each log entry.
+* There should at least one production ready implementation of the selected API which could produce JSON formatted logs.
+* It should be possible to produce klog compatible logs from the selected API which could be used during the migration to a new API.
+
+Based on this criteria https://github.com/go-logr/logr log interface has been selected as a logging API for Kubernetes.
```

Comment:
We will add preparation of the verbosity level guide as one of the requirements for migration.

Reply:
@brancz This is one of the most contentious points in the entirety of the logging space, and has been for years and years.  It's pretty much religious.

Since I am responsible for logr, here's my rationale.  In my experience:

* Named logging levels are no more or less consistent than numeric levels.  Within some scope (e.g. a package) they may be consistent but what makes a "warning" different from an "info", or a "trace" from a "debug" is ill-defined.

* Named levels are more confusing because there are relatively few of them, so their definitions get "bent" to fit the needs.

* Remapping or offsetting log levels is often useful when passing loggers to external libs that don't follow your preferred convention for levels.  Named levels implicitly get numeric-like values.

* Most projects end up adopting conventions like "don't use WARNING" but those don't apply outside of a projects (e.g. Kubernetes vendors most of the go code on github).

* A simple rule - higher levels log more - is easy to comprehend.  Setting conventions like levels "higher than 10 mean debugging/tracing" fit within that and allow package owners to instrument their code for incrementally chatty tracing.

Given these, I opted for 2 key design decisions in logr.  1) Error and Info, no Warning.  2) numeric levels.  Nothing says you can't define consts for levels within your own package or project.

logr is still "beta" and there are some open questions on it.  I'd be very happy to have more input and discussion on it.  My not-so-subtle goal is to use kubernetes to help establish a de facto logging interface for the Go ecosystem.  I like logr, but if the goal requires API changes to be achieved, I am not above that :)

Reply:
> It's pretty much religious.

Agreed, which is why I'm not trying to force my opinion on the project, I don't think it's helpful.

> My not-so-subtle goal is to use kubernetes to help establish a de facto logging interface for the Go ecosystem.

Noble goal, but I think there are too many other established ones that this would actually happen. I think judging by what we want to achieve and our existing practices and code, I think logr is the right choice for Kubernetes, I'm not entirely sure whether the rest of the ecosystem agrees though :) .